### PR TITLE
Add supplementary information to transpiler module

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
Add supplementary information that was originally in the tutorials to the transpiler module.

### CODE REVIEW

File: qiskit/transpiler/__init__.py [hunk 1]
```diff
@@ -19,8 +19,403 @@
...
+
+      <br>
+
+
+Transpiler API
+==============
+
 Pass Management
-===============
+---------------
```
Comment (ajavadia) [inline qiskit/transpiler/__init__.py:418]: Can we add `transpiler.preset_passmanagers` to the list of documented modules? There is helpful information in there about what exactly each level does, but right now it is not picked up by sphinx.

File: qiskit/transpiler/__init__.py [hunk 2]
```diff
@@ -19,8 +19,403 @@
...
+      from qiskit.test.mock import FakeVigo
+      backend = FakeVigo()
+
+   .. jupyter-execute::
+
+      backend.configuration().basis_gates
+
+   We see that IBM Quantum devices support five native gates: four single-qubit gates
+   [`u1`, `u2`, `u3`, and `id`] and one two-qubit entangling gate `cx`.  In addition, the
+   devices support qubit measurements (otherwise we can not read out an answer!).
```
Comment (ajavadia) [inline qiskit/transpiler/__init__.py:74]: This whole section that goes into detail about u1,u2,u3,cx is a bit orthogonal to what the transpiler is. Those gates are already documented, here I think we should go into detail about different transpiler passes. Besides, these u gates are going to go obsolete soon: https://github.com/Qiskit/qiskit-terra/issues/4106

File: qiskit/transpiler/__init__.py [hunk 3]
```diff
@@ -19,8 +19,403 @@
...
+
+Most circuits must undergo a series of transformations that make them compatible with
+a given target device, and optimize them to reduce the effects of noise on the
+resulting outcomes.  Rewriting quantum circuits to match hardware constraints and
+optimizing for performance can be far from trivial.  The flow of logic in the rewriting
+tool chain need not be linear, and can often have iterative sub-loops, conditional
+branches, and other complex behaviors.  That being said, the basic building blocks
+follow the structure given below:
+
+.. image:: /source_images/transpiling_core_steps.png
```
Comment (ajavadia) [inline qiskit/transpiler/__init__.py:36]: Can we change this to something more like:
```
Virtual circuit optimization
Decompose gates acting on 3 or more qubits
Placement on physical qubits ("layout")
Routing on restricted topology
Translate to basis
Physical circuit optimization
```

File: qiskit/transpiler/__init__.py (hunk 1 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py:418]: Sure.

File: qiskit/transpiler/__init__.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py:74]: Ok will cut out. Not sure how to reconcile the desire to talk about transpiler passes here with exposing them into the docs.  Surely the passes module should discuss this.  I can however link to the docs there and then the users would know exactly what is going on when someone gets around to making those docs.

File: qiskit/transpiler/__init__.py (hunk 3 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py:36]: ok

File: qiskit/transpiler/__init__.py (hunk 1 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py]: This is already at the top-level of the documentation.

File: qiskit/transpiler/__init__.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py]: Passes are also already at top-level in the docs.

**Comment by nonhermitian** [comment]
@ajavadia the requested changes have been made.

File: qiskit/transpiler/__init__.py [hunk 4]
```diff
@@ -19,8 +19,364 @@
...
+
+   - **Trivial layout**: Map virtual qubits to the same numbered physical qubit on the device,
+     i.e. `[0,1,2,3,4]` -> `[0,1,2,3,4]` (default in `optimization_level=0`).
+
+   - **Dense layout**: Find the sub-graph of the device with same number of qubits as the circuit
+     with the greatest connectivity (default in `optimization_level=1`).
+
+   - **Noise adaptive layout**: Uses the noise properties of the device, in concert with the
+     circuit properties, to generate the layout with the best noise properties (default in
+     `optimization_level=2` and `optimization_level=3`).
```
Comment (mtreinish) [inline qiskit/transpiler/__init__.py:185]: This isn't quite true anymore. Dense layout is the default in level 2 and 3 now that it's noise aware:

https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/preset_passmanagers/level2.py#L80
https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/preset_passmanagers/level3.py#L84

File: qiskit/transpiler/__init__.py [hunk 5]
```diff
@@ -19,8 +19,364 @@
...
+
+   In order to implement a CNOT gate between qubits in a quantum circuit that are not directly
+   connected on a quantum device one or more SWAP gates must be inserted into the circuit to
+   move the qubit states around until they are adjacent on the device gate map.  Each SWAP
+   gate is decomposed into three CNOT gates on the IBM Quantum devices, and represents an
+   expensive and noisy operation to perform.  Thus, finding the minimum number of SWAP gates
+   needed to map a circuit onto a given device, is an important step (if not the most important)
+   in the whole execution process.
+
+   However, as with many important things in life, finding the optimal SWAP mapping is hard.
```
Comment (mtreinish) [inline qiskit/transpiler/__init__.py:269]: Heh, this is a bit philosophical for documentation :)

Comment (nonhermitian) [reply qiskit/transpiler/__init__.py:269]: Well it is supplementary material so all bets are off.

File: qiskit/transpiler/__init__.py (hunk 4 above)
Comment (nonhermitian) [reply qiskit/transpiler/__init__.py:185]: Ahh another change.  Ok.

**Comment by nonhermitian** [comment]
Blocked by #4241

**Comment by 1ucian0** [comment]
#4241 was closed. Is this ready?

**Comment by nonhermitian** [comment]
Been ready for a while now.

**Comment by 1ucian0** [comment]
requesting approval from @ajavadia and @mtreinish in that case. Thanks!

**Pull request merged**